### ISSUE TITLE
Fallback is being called if user authentication fails on LDAP

### ISSUE THREAD

**Issue by pauls6832** [Adldap2/Adldap2-Laravel#422]
- Laravel Version: 5.5
- Adldap2-Laravel Version: 3.0.5
- PHP Version: 7.0

### Description:

I've noticed that if a user fails to login to LDAP, e.g. if password is wrong or if user is de-activated, then the system uses the fallback authentication.

Shouldn't the correct behaviour be to only call the fallback if it cannot find the user in LDAP? If the user exists in LDAP but authentication fails then it should not revert to the fallback.

There seems to be two problems related to the fallback:

1. Users whose passwords have been reset can still login using old password. LDAP authentication fails so system resorts to database which still has the old password.

2. The system never updates the database record with the new password because the model only gets written to the database when LDAP authentication is successful. If authentication fails the model is never saved and the database keeps the old password. It only changes if the user successfully logs into LDAP.

I might be miss-understanding the purpose of fallback, but this seems wrong to me.

I've noticed a few reports of people still being able to login with old passwords. I suspect this is the cause of those problems, not LDAP caching passwords but the database still storing old password.

### Steps To Reproduce:

1. Create user in LDAP

2. Login successfully, database record is created for user.

3. Change password in LDAP

4. User can still login with old password. Old password remains in database until user logs in with new password. NOTE: This is not a caching problem.

5. If user logs in with new password then database is updated with new password and user can no longer use the old password.

### Possible Fix

I think the problem is in the DatabaseUserProvider validateCredentials() method. (comments removed)

```
public function validateCredentials(Authenticatable $model, array $credentials)
{
    if (
        $this->user instanceof User &&
        $this->getResolver()->authenticate($this->user, $credentials)
    ) {
        $this->handleAuthenticatedWithCredentials($this->user, $model);
        if ($this->newValidator($this->getRules($this->user, $model))->passes()) {
            if ($this->isSyncingPasswords()) {
                $password = $credentials['password'];
            } else {
                $password = str_random();
            }
            $model->password = $model->hasSetMutator('password') ? $password : bcrypt($password);
            $model->save();
            if ($this->isBindingUserToModel($model)) {
                $model->setLdapUser($this->user);
            }
            return true;
        }
    }
    if ($this->isFallingBack() && $model->exists) {
        return $this->fallback->validateCredentials($model, $credentials);
    }
     return false;
}
```

The problem seems to be in the first IF statement:

```
if (
    $this->user instanceof User &&
    $this->getResolver()->authenticate($this->user, $credentials)
) {
```

If these are separated, and the model always saved as long as it exists, regardless of whether the user authenticates, then the database will be updated with the correct credentials. Then if authentication fails we should return false, not revert to the fallback.

I thought this might work better:

```
public function validateCredentials(Authenticatable $model, array $credentials)
{
    if ($this->user instanceof User) {
        // If user exists, save with updated data from LDAP
        $model->save();

        // Authenticate the user
        if ($this->getResolver()->authenticate($this->user, $credentials)) {
            ...
        }

        // Return false if authentication fails
        return false;
    }

    // If user not found in LDAP revert to fallback
    if ($this->isFallingBack() && $model->exists) {
        return $this->fallback->validateCredentials($model, $credentials);
    }

    return false;
}
```

**Comment by stevebauman** [comment]
Hi @pauls6832,

This is currently the intended usage of the fallback option, however it is definitely up for debate, and I have no issues changing the current way it's implemented.

Though, I have implemented it this way purposely so if for some reason there is no LDAP connectivity (for example if your AD server is down), then users can still login with their old password to be able to access your application.

I felt like this scenario was helpful for 'just in case' events where this may occur.

I would love for some more user feedback on this so I can change the behavior if this is what the majority would prefer.

Does anyone have any thoughts on this?

**Comment by pauls6832** [comment]
In the current version (v3.0.5) it throws an exception anyway if LDAP is down, so the system never gets to the fallback. I understand you are changing that in the next release (v4?). 

This change still allows it to work the way you want. If LDAP is down then the first check fails, the system won't find the user in LDAP and will then use the fallback. I've done a quick test and it works, might need more thorough testing though. As long as $user doesn't return an instance of User if LDAP fails it will work.

The main advantage is that a) LDAP is the primary source of authentication, if LDAP is up and authentication fails then login fails, if LDAP is down then fallback is called, b) the database will always have the last updated information from LDAP, no "caching" of the old password.

This last is important for me not just because of the "old" password problem but because I also have an "active" field in the database, users can only login if they are active. If this value is changed on LDAP it needs updating in the database but with current system that may not happen if user can't login, e.g. if status is changed to Deactivated in LDAP the LDAP login fails, but then the system uses fallback and local login works because the status is still active in the database. There is another problem I'm having related to this, I'll have to log another ticket for that.

**Comment by pauls6832** [comment]
Sorry, just to clarify. The testing I did was based on the system not throwing an exception if LDAP is down. I just commented out the Exception in Guard.php. It won't work with current 3.0.5 but will work when next version comes out where you don't throw the exception.

**Comment by 36864** [comment]
The term fallback can mean either "use this if ldap is down" or "use this if the user isnt found on the ldap server".

The one you prefer depends on whether you prefer security or availability.

This should be something that users can set through the configuration files,  I think. Personally I would always prefer to deny authentication if LDAP is down (and that works right now for me, even if the exception isn't thrown because I simply don't store user passwords locally), but I can see use cases where you might want your stuff to keep working through some ldap downtime.

Maybe if there was a maximum staleness value that you could set, we could have a bit of a compromise between both scenarios. For example with a staleness value of 2 hours, if your LDAP goes down, users can still log in as long as their information in the database is less than 2 hours old (read: they have logged in in the past 2 hours OR there could be an automatic data sync refreshing user data).

Setting this value to a relatively low value would let you keep working through minor service outages while not having to worry too much about old passwords.

**Comment by pauls6832** [comment]
The original issue in this ticket I think is a security problem. The secondary one of whether to raise an Exception when LDAP is down or use fallback is just a matter of preference. I see there's plenty of discussions about pros and cons of that, maybe a config option is easiest way sort that out.

The reason I think the original is a security issue is because when LDAP IS working but authentication fails for a valid user then instead of login failing the system uses the fallback option. The fallback reads from the database and the database still has the old password in it. Fallback shouldn't be called in this case (where user exists, but password is wrong), it should only be called if the user does not exist in LDAP.

As an example, lets say a user's password is compromised by a hacker. The user changes their password in LDAP, but doesn't login to the PHP application immediately. The hacker will still be able to login to the PHP application using the compromised password. They will login using the old password, LDAP authentication will fail because the password is wrong, BUT the system will then use fallback and the database still has the old, compromised password in it.

**Comment by 36864** [comment]
I misunderstood the problem at first. I agree completely, this needs to be changed.

**Comment by stevebauman** [comment]
Thanks guys! I'll patch this immediately on the current release as well as master.

Will post again shortly once this has been patched so I can go through your suggestions further.

**Comment by stevebauman** [comment]
Thanks again, I've patched this in [v3.0.6](https://github.com/Adldap2/Adldap2-Laravel/commit/3cd8cca52e2949bb92e6c09bbe91623dff86f4db) (as well in [master](https://github.com/Adldap2/Adldap2-Laravel/commit/d3a29edfa9582b019b9ae44eb1521a007b0b330d)) so you can update your applications now.

If any of you experience any further issues or want to make other suggestions just leave a comment or create another issue. :)
